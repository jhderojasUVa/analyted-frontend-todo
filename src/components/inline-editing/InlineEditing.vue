<script>
export default {
    name: 'InlineEditing',
    props: {
        data: String
    },
    mounted() {
        // set the focus on the description
        // so when moving out we emit the changes
        this.$refs.description.focus();
    },
    methods: {
        updateText() {
            // emit a change on the description
            this.$emit('descriptionChanged', this.data);
        }
    }
}
</script>

<template>
    <input ref="description" type="text" v-model="data" v-on:blur="updateText" size="20" maxlength="50" />
</template>

<style scoped>
    input {
        padding: 0.5em;
        border-radius: 3px;
        border: 1px solid rgba(0, 0, 0, 0.5);
    }
</style>
