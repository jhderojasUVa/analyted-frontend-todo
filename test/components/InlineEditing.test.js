import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import InlineEditing from '../../src/components/inline-editing/InlineEditing.vue';

describe('InlineEditing' ,() => {
    it('renders', () => {
        const data = 'foo';

        const wrapper = mount(InlineEditing, { props: { data } });

        expect(wrapper).toMatchSnapshot();
    });
    it('changes the content, works', async () => {
        const description = 'foo';
        const newDescription = 'bar';

        const wrapper = mount(InlineEditing, { props: { description } });

        // search for input
        const input = wrapper.find('input');

        // set new value
        await input.setValue(newDescription);

        // set focus
        await input.trigger('focus');
        
        wrapper.vm.$nextTick();

        // loose focus (here it emits)
        await input.trigger('focusout');

        wrapper.vm.$nextTick();

        // check
        expect(wrapper.find('input').element.value).toBe(newDescription);
        expect(wrapper.vm.data).toBe(newDescription);
        expect(wrapper.emitted().focusout).not.toBe(undefined);
    });
});
