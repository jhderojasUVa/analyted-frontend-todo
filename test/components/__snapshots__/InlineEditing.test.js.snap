// Vitest Snapshot v1

exports[`InlineEditing > renders 1`] = `
VueWrapper {
  "__app": {
    "_component": {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <input
        data-v-b406172e=""
        maxlength="50"
        size="20"
        type="text"
      />
    </div>,
    "_context": {
      "app": [Circular],
      "components": {
        "transition": {
          "name": "transition",
        },
        "transition-group": {
          "name": "transition-group",
        },
      },
      "config": {
        "compilerOptions": {},
        "errorHandler": undefined,
        "globalProperties": {},
        "isNativeTag": [Function],
        "optionMergeStrategies": {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": {},
      "emitsCache": WeakMap {},
      "mixins": [
        {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": {},
      "reload": [Function],
    },
    "_instance": {
      "a": null,
      "accessCache": {},
      "appContext": {
        "app": [Circular],
        "components": {
          "transition": {
            "name": "transition",
          },
          "transition-group": {
            "name": "transition-group",
          },
        },
        "config": {
          "compilerOptions": {},
          "errorHandler": undefined,
          "globalProperties": {},
          "isNativeTag": [Function],
          "optionMergeStrategies": {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": {},
        "emitsCache": WeakMap {},
        "mixins": [
          {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": {},
      "da": null,
      "data": {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": [
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
        ],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": {},
      "propsDefaults": {},
      "propsOptions": [
        {},
        [],
      ],
      "provides": {},
      "proxy": {},
      "refs": {
        "VTU_COMPONENT": {
          "data": "foo",
          "hasOwnProperty": [Function],
          "updateText": [Function],
        },
      },
      "render": [Function],
      "renderCache": [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": [],
        "effects": [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": {},
      "slots": {},
      "sp": null,
      "subTree": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": {
          "a": null,
          "accessCache": {
            "@@__IMMUTABLE_ITERABLE__@@": 0,
            "@@__IMMUTABLE_RECORD__@@": 0,
            "_isMockFunction": 0,
            "constructor": 0,
            "data": 3,
            "hasOwnProperty": 4,
            "toJSON": 0,
            "updateText": 4,
            Symbol(Symbol.toStringTag): 0,
          },
          "appContext": {
            "app": [Circular],
            "components": {
              "transition": {
                "name": "transition",
              },
              "transition-group": {
                "name": "transition-group",
              },
            },
            "config": {
              "compilerOptions": {},
              "errorHandler": undefined,
              "globalProperties": {},
              "isNativeTag": [Function],
              "optionMergeStrategies": {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": {},
            "emitsCache": WeakMap {},
            "mixins": [
              {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": {},
          "ctx": {
            "data": "foo",
            "hasOwnProperty": [Function],
            "updateText": [Function],
          },
          "da": null,
          "data": {},
          "directives": null,
          "ec": null,
          "effect": ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
          "emit": [Function],
          "emitsOptions": null,
          "emitted": null,
          "exposeProxy": null,
          "exposed": null,
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": [
            [Function],
          ],
          "next": null,
          "parent": [Circular],
          "props": {
            "data": "foo",
          },
          "propsDefaults": {},
          "propsOptions": [
            {
              "data": {
                "0": false,
                "1": true,
                "type": [Function],
              },
            },
            [],
          ],
          "provides": {},
          "proxy": {
            "data": "foo",
            "hasOwnProperty": [Function],
            "updateText": [Function],
          },
          "refs": {
            "description": <input
              data-v-b406172e=""
              maxlength="50"
              size="20"
              type="text"
            />,
          },
          "render": [Function],
          "renderCache": [
            [Function],
            [Function],
          ],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": [],
            "effects": [
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": [
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "parent": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": null,
          "setupState": {},
          "slots": {},
          "sp": null,
          "subTree": {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": null,
            "component": null,
            "dirs": [
              {
                "arg": undefined,
                "dir": {
                  "beforeUpdate": [Function],
                  "created": [Function],
                  "mounted": [Function],
                },
                "instance": {
                  "data": "foo",
                  "hasOwnProperty": [Function],
                  "updateText": [Function],
                },
                "modifiers": {},
                "oldValue": undefined,
                "value": "foo",
              },
            ],
            "dynamicChildren": [],
            "dynamicProps": null,
            "el": <input
              data-v-b406172e=""
              maxlength="50"
              size="20"
              type="text"
            />,
            "key": null,
            "patchFlag": 544,
            "props": {
              "maxlength": "50",
              "onBlur": [Function],
              "onUpdate:modelValue": [Function],
              "ref": "description",
              "size": "20",
              "type": "text",
            },
            "ref": {
              "f": false,
              "i": [Circular],
              "k": undefined,
              "r": "description",
            },
            "scopeId": "data-v-b406172e",
            "shapeFlag": 1,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "input",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": {
            "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/inline-editing/InlineEditing.vue",
            "__scopeId": "data-v-b406172e",
            "components": {},
            "methods": {
              "updateText": [Function],
            },
            "mounted": [Function],
            "name": "InlineEditing",
            "props": {
              "data": [Function],
            },
            "render": [Function],
          },
          "u": null,
          "uid": 1,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <input
          data-v-b406172e=""
          maxlength="50"
          size="20"
          type="text"
        />,
        "key": null,
        "patchFlag": 0,
        "props": {
          "data": "foo",
          "ref": "VTU_COMPONENT",
        },
        "ref": {
          "f": false,
          "i": [Circular],
          "k": undefined,
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/inline-editing/InlineEditing.vue",
          "__scopeId": "data-v-b406172e",
          "components": {},
          "methods": {
            "updateText": [Function],
          },
          "mounted": [Function],
          "name": "InlineEditing",
          "props": {
            "data": [Function],
          },
          "render": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 0,
      "um": null,
      "update": [Function],
      "vnode": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": {
          "app": [Circular],
          "components": {
            "transition": {
              "name": "transition",
            },
            "transition-group": {
              "name": "transition-group",
            },
          },
          "config": {
            "compilerOptions": {},
            "errorHandler": undefined,
            "globalProperties": {},
            "isNativeTag": [Function],
            "optionMergeStrategies": {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": {},
          "emitsCache": WeakMap {},
          "mixins": [
            {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <input
          data-v-b406172e=""
          maxlength="50"
          size="20"
          type="text"
        />,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": {
      "compilerOptions": {},
      "errorHandler": undefined,
      "globalProperties": {},
      "isNativeTag": [Function],
      "optionMergeStrategies": {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.37",
  },
  "__setProps": [Function],
  "componentVM": {
    "data": "foo",
    "hasOwnProperty": [Function],
    "updateText": [Function],
  },
  "isDisabled": [Function],
  "rootVM": {},
  "wrapperElement": <input
    data-v-b406172e=""
    maxlength="50"
    size="20"
    type="text"
  />,
}
`;
